src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/InteropServices/DisableRuntimeMarshalling.cs

```diff
@@ -0,0 +1,354 @@
+                IInvocationOperation invocation = (IInvocationOperation)context.Operation;
+
+                if (_marshalMethods.Contains(invocation.TargetMethod.ConstructedFrom))
+                {
+                    bool canTransformToDisabledMarshallingEquivalent = CanTransformToDisabledMarshallingEquivalent(invocation);
+                    context.ReportDiagnostic(invocation.CreateDiagnostic(
+                        MethodUsesRuntimeMarshallingEvenWhenMarshallingDisabled,
+                        ImmutableDictionary.Create<string, string?>().Add(CanConvertToDisabledMarshallingEquivalentKey, canTransformToDisabledMarshallingEquivalent ? "true" : null),
+                        invocation.TargetMethod.ToDisplayString()));
+                }
+
+                static bool CanTransformToDisabledMarshallingEquivalent(IInvocationOperation invocation)
+                {
+                    return invocation.TargetMethod.Name switch
+                    {
+                        "OffsetOf" => false,
+                        "SizeOf" => invocation.TargetMethod.IsGenericMethod || (invocation.Arguments.Length > 0 && invocation.Arguments[0].Value is ITypeOfOperation { TypeOperand.IsUnmanagedType: true }),
+                        "StructureToPtr" => invocation.Arguments.Length > 0 && invocation.Arguments[0].Value is { Type.IsUnmanagedType: true },
+                        "PtrToStructure" => invocation.Type is not null,
+                        _ => false
+                    };
+                }
+            }
+
+            public void AnalyzeFunctionPointerCall(OperationAnalysisContext context)
+            {
+                var functionPointerInvocation = IFunctionPointerInvocationOperationWrapper.FromOperation(context.Operation);
+
+                if (functionPointerInvocation.GetFunctionPointerSignature().CallingConvention() == System.Reflection.Metadata.SignatureCallingConvention.Default)
+                {
+                    return;
+                }
+
+                AnalyzeMethodSignature(context.ReportDiagnostic, functionPointerInvocation.GetFunctionPointerSignature(), ImmutableArray.Create(functionPointerInvocation.WrappedOperation.Syntax.GetLocation()));
+            }
+
+            public void AnalyzeType(SymbolAnalysisContext context)
+            {
+                Debug.Assert(_unmanagedFunctionPointerAttribute is not null);
+                INamedTypeSymbol type = (INamedTypeSymbol)context.Symbol;
+                if (type.TypeKind != TypeKind.Delegate || !type.HasAttribute(_unmanagedFunctionPointerAttribute))
+                {
+                    return;
+                }
+
+                AnalyzeMethodSignature(context.ReportDiagnostic, type.DelegateInvokeMethod);
+            }
+
+            public void AnalyzeMethod(SymbolAnalysisContext context)
+            {
+                IMethodSymbol method = (IMethodSymbol)context.Symbol;
+
+                DllImportData? dllImportData = method.GetDllImportData();
```

Comment:
I'm surprised this didn't cause a crash due to:

https://github.com/dotnet/roslyn/blob/7e2966e180eef601cb0fc4a029d4f2093d22b639/src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs#L829

Also surprised on roslyn's side there is an exception thrown in this code path instead of simply returning null. cc @333fred

Reply:
We don't go down this path for function pointers. We go from `AnalyzeFunctionPointerCall` to `AnalyzeMethodSignature` directly.

Reply:
`FunctionPointerMethodSymbol` is a *method* symbol. Shouldn't we get callbacks for it from `                context.RegisterSymbolAction(AnalyzeMethod, SymbolKind.Method);`? then we meet `method.GetDllImportData()` unconditionally. I know this is likely not happening, otherwise a test would have crashed. But I don't yet see the reason

Reply:
I think we don't get callbacks for it for the same reason that we don't get callbacks for `SymbolKind.Method` for local functions (which I just added code to handle in my most recent commit). It seems that symbols that are introduced due to an `IOperation` are exposed through that operation in `RegisterOperationAction` instead of through `RegisterSymbolAction`.

Reply:
Good to know! Thanks!